src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java

```diff
@@ -191,11 +252,24 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
 
         List<ProgrammingSubmission> submissions = programmingExerciseService.notifyChangedTestCases(exerciseId, requestBody);
 
-        // notify users via websocket.
+        notifyUserTriggerBuildForNewSubmissions(submissions);
+
+        return ResponseEntity.ok().build();
+    }
+
+    private void notifyUserTriggerBuildForNewSubmissions(Collection<ProgrammingSubmission> submissions) {
         for (ProgrammingSubmission submission : submissions) {
-            messagingTemplate.convertAndSend("/topic/participation/" + submission.getParticipation().getId() + "/newSubmission", submission);
+            notifyUserAndTriggerBuildForNewSubmission(submission);
         }
+    }
 
-        return ResponseEntity.ok().build();
+    /**
+     * Sends a websocket message to the user about the new submission and triggers a build on the CI system.
+     *
+     * @param submission ProgrammingSubmission that was just created.
+     */
+    private void notifyUserAndTriggerBuildForNewSubmission(ProgrammingSubmission submission) {
+        messagingTemplate.convertAndSend("/topic/participation/" + submission.getParticipation().getId() + Constants.PROGRAMMING_SUBMISSION_TOPIC, submission);
```

Comment:
This looks a bit strange for me. Why do we first send the new submission to the user and then trigger the build on the CI server? Should it not be the other way round?

What happens if the CI server does not respond or throws an exception? See e.g. https://docs.atlassian.com/atlassian-bamboo/REST/5.5.0/#d2e823

Reply:
You are right, this is strange.
The handling of informing the user vs. triggering the build is inconsistent.
I adjusted it for most parts to send an error object to the user in case the communication to Bamboo fails, though there is a case in the service for `resumeExercise` that is not easy to handle without restructuring, so I added a TODO.